pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -299,47 +293,97 @@ func (s *Service) sendMsgEvents(msg *service.StateMsg) {
 	go func() {
 		for msg := range s.callbackChannel {
-			// TODO handle error in callback - https://github.com/hyperledger/aries-framework-go/issues/242
-			if err := s.process(msg); err != nil {
-				// TODO handle error
-				logger.Errorf(err.Error())
+			if msg.Err != nil {
+				// TODO https://github.com/hyperledger/aries-framework-go/issues/438 Cleanup/Update data on Stop Event Action
+				logger.Errorf("client action event processing failed - msgID:%s error:%s", msg.ID, msg.Err)
+				err := s.processFailure(msg.ID, msg.Err)
+				if err != nil {
+					logger.Errorf("process callback failure: %s", err)
+				}
+				continue
+			}
+
+			// TODO https://github.com/hyperledger/aries-framework-go/issues/242 - retry logic
+
+			if err := s.process(msg.ID); err != nil {
+				procErr := s.processFailure(msg.ID, err)
+				if procErr != nil {
+					logger.Errorf("process callback failure: %s", procErr)
+				}
 			}
 		}
 	}()
 }
 
+func (s *Service) processFailure(id string, processErr error) error {
+	// get the transient data
+	data, err := s.getTransientEventData(id)
+	if err != nil {
+		return fmt.Errorf("unable to fetch the event transient data: %w", err)
+	}
+
+	err = s.abandon(data.ThreadID, data.Msg, processErr)
+	if err != nil {
+		return fmt.Errorf("process failure: %w", err)
+	}
+
+	return nil
+}
+
+// abandon updates the state to abandoned and trigger failure event.
+func (s *Service) abandon(thid string, msg *service.DIDCommMsg, processErr error) error {
+	// update the state to abandoned
+	err := s.update(thid, &abandoned{})
+	if err != nil {
+		return fmt.Errorf("unable to update the state to abandoned: %w", err)
+	}
+
+	// send the message event
+	s.sendMsgEvents(&service.StateMsg{
+		Type:       service.PostState,
+		Msg:        msg,
+		StateID:    (&abandoned{}).Name(),
+		Properties: createErrorEventProperties(thid, "", processErr),
+	})
+
+	return nil
+}
+
 func (s *Service) processCallback(id string, err error) {
 	// pass the callback data to internal channel. This is created to unblock consumer go routine and wrap the callback
 	// channel internally.
 	s.callbackChannel <- didCommChMessage{ID: id, Err: err}
 }
 
 // processCallback processes the callback events.
-func (s *Service) process(msg didCommChMessage) error {
-	if msg.Err != nil {
-		// TODO https://github.com/hyperledger/aries-framework-go/issues/438 Cleanup/Update data on Stop Event Action
-		logger.Errorf("client action event processing failed - msgID:%s error:%s", msg.ID, msg.Err)
-		return nil
+func (s *Service) process(id string) error {
+	data, err := s.getTransientEventData(id)
+	if err != nil {
+		return fmt.Errorf("unable to fetch the event transient data: %w", err)
 	}
 
-	// fetch the record
-	jsonDoc, err := s.store.Get(msg.ID)
+	// continue the processing
+	err = s.handle(data)
 	if err != nil {
-		return fmt.Errorf("document for the id doesn't exists in the database: %w", err)
+		return fmt.Errorf("processing of the message failed: %w", err)
```

Comment:
do we need to have every wrapping error string say failed? :)

keep it succinct.

Reply:
[Removed](https://github.com/hyperledger/aries-framework-go/compare/807adcf62e0ac91e48d81a5b15a62d91e073bc78..0360003cb9f51624a1493097c2398be2113925f0#diff-f761c91e753eba1c42a68076df06ce12R298) failed word :)